File: app/src/main/java/com/water/intentsample/IntentActivity.java
package com.water.intentsample;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.ContactsContract;
import android.provider.MediaStore;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.FileProvider;
import androidx.databinding.DataBindingUtil;

import com.water.intentsample.databinding.ActivityIntentBinding;
import com.water.utilities.PermissionHelper;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;

public class IntentActivity extends AppCompatActivity {

    private Context _context;

    String[] _permissions = {
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    };

    private final String tag = this.getClass().getSimpleName();

    private ActivityIntentBinding binding;

    /**
     * 注册返回数据回调
     */
    private final ActivityResultLauncher<Intent> GetReturnString = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == RESULT_OK) {
                    assert result.getData() != null;
                    Bundle MarsBundle = result.getData().getExtras();
                    String msg = MarsBundle.getString(MarsActivity.MarsResultName);
                    Toast.makeText(this, msg, Toast.LENGTH_LONG).show();
                }
            });

    /**
     * 注册TakePhoto回调
     */
    private final ActivityResultLauncher<Intent> TakePhoto = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == RESULT_OK) {
                    // 取出照片数据
                    assert result.getData() != null;
                    Bundle extras = result.getData().getExtras();
                    Bitmap bmp = (Bitmap) extras.get("data");
                    // 根据旋转角度，生成旋转矩阵
                    Matrix matrix = new Matrix();
                    matrix.postRotate(90);

                    try {
                        // 将原始图片按照旋转矩阵进行旋转，并得到新的图片
                        Bitmap bitmap = Bitmap.createBitmap(bmp, 0, 0, bmp.getWidth(),
                                bmp.getHeight(), matrix, true);
                        binding.ivPhoto.setImageBitmap(bitmap);
                        bmp.recycle();
                    } catch (OutOfMemoryError e) {
                        Log.e(tag, "OutOfMemoryError in IntentActivity", e);
                    }
                }
            });

    @SuppressLint("QueryPermissionsNeeded")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_intent);

        _context = this.getBaseContext();

        int columnCount = binding.glContainer.getColumnCount();

        DisplayMetrics dm = new DisplayMetrics();
        this.getWindowManager().getDefaultDisplay().getMetrics(dm);
        int screenWidth = dm.widthPixels;

        for (int i = 0; i < binding.glContainer.getChildCount(); i++) {
            Button btn = (Button) binding.glContainer.getChildAt(i);
            btn.setWidth(screenWidth / columnCount);
        }

        // 返回首页
        binding.btnHome.setOnClickListener(v -> {
            Intent home = new Intent();
            //为Intent设置Action和Category属性
            home.setAction(Intent.ACTION_MAIN);
            home.addCategory(Intent.CATEGORY_HOME);

            //使用resolveActivity进行检查，这是十分必要的
            if(home.resolveActivity(getPackageManager())!=null){
                startActivity(home);
            }else{
                Toast.makeText(_context,"未找到对应应用",Toast.LENGTH_SHORT).show();
            }
        });

        // 给移动客服10086拨打电话
        binding.btnTel.setOnClickListener(v -> {
            Intent intent = new Intent(Intent.ACTION_DIAL);
            intent.setData(Uri.parse("tel:10086"));

            if(intent.resolveActivity(getPackageManager())!=null){
                startActivity(intent);
            }else{
                Toast.makeText(_context,"未找到对应应用",Toast.LENGTH_SHORT).show();
            }
        });

        // 给10086发送内容为“Hello”的短信
        binding.btnSms.setOnClickListener(v -> {
            // 关于smsto的内容，请参考下列链接
            // https://developer.android.com/guide/components/intents-common?hl=zh-cn
            Uri us = Uri.parse("smsto:10086");
            Intent intent = new Intent(Intent.ACTION_SENDTO, us);
            intent.putExtra("sms_body", "Hello");

            if (intent.resolveActivity(getPackageManager()) != null) {
                startActivity(intent);
            } else {
                Toast.makeText(_context, "未找到对应应用", Toast.LENGTH_SHORT).show();
            }
        });

        // 打开浏览器
        binding.btnUrl.setOnClickListener(v -> {
            Uri uu = Uri.parse("http://www.baidu.com");
            Intent intent  = new Intent(Intent.ACTION_VIEW, uu);

            if(intent.resolveActivity(getPackageManager())!=null){
                startActivity(intent);
            }else{
                Toast.makeText(_context,"未找到对应应用",Toast.LENGTH_SHORT).show();
            }
        });

        binding.btnWireless.setOnClickListener(v -> {
            Intent intent = new Intent(android.provider.Settings.ACTION_WIRELESS_SETTINGS);

            if(intent.resolveActivity(getPackageManager())!=null){
                startActivity(intent);
            }else{
                Toast.makeText(_context,"未找到对应应用",Toast.LENGTH_SHORT).show();
            }
        });

        binding.btnContent.setOnClickListener(v -> {
            Intent intent = new Intent();
            intent.setAction(Intent.ACTION_VIEW);
            intent.setData(ContactsContract.Contacts.CONTENT_URI);

            if(intent.resolveActivity(getPackageManager())!=null){
                startActivity(intent);
            }else{
                Toast.makeText(_context,"未找到对应应用",Toast.LENGTH_SHORT).show();
            }
        });

        binding.btnExplicitStart.setOnClickListener(v -> {
            Intent intent = new Intent(IntentActivity.this, FirstActivity.class);

            if(intent.resolveActivity(getPackageManager())!=null){
                startActivity(intent);
            }else{
                Toast.makeText(_context,"未找到对应应用",Toast.LENGTH_SHORT).show();
            }
        });

        binding.btnReturnData.setOnClickListener(v -> {
            Intent mars = new Intent(IntentActivity.this,
                    MarsActivity.class);
            String content = "地球来的消息:我是来自地球上的Tom，火星的朋友你好！";
            mars.putExtra("FromEarth", content);

            GetReturnString.launch(mars);
        });

        binding.btnComplexData.setOnClickListener(v -> {
            Intent intent = new Intent(IntentActivity.this, ComplexDataActivity.class);

            // 传递基本数据类型或String类型数组
            Bundle bd = new Bundle();
            bd.putStringArray("StringArray", new String[]{"请求","全球"});

            // 传递基本数据类型或String类型集合
            intent.putStringArrayListExtra("StringList", new ArrayList<String>(){{
                add("请求");
                add("全球");
            }});

            // 传递Bitmap
            try {
                InputStream in = getResources().getAssets().open("car.png");
                Bitmap bmp = BitmapFactory.decodeStream(in);
                bd.putParcelable("bitmap", bmp);
            } catch (IOException e) {
                e.printStackTrace();
            }

            intent.putExtras(bd);

            if(intent.resolveActivity(getPackageManager()) != null){
                startActivity(intent);
            }else{
                Toast.makeText(_context,"未找到对应应用",Toast.LENGTH_SHORT).show();
            }

        });

        // 相当于发送带附件的短信
        binding.btnImageSms.setOnClickListener(v -> {
            Intent iis=new Intent();
            iis.setAction(Intent.ACTION_SEND);
            iis.putExtra("address","10086"); // 邮件地址
            iis.putExtra("sms_body","Hello"); // 邮件内容
            File fsms = new File(Environment.getExternalStorageDirectory(),
                    "/Pictures/kb.png");
            Uri uri;

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                uri= FileProvider.getUriForFile(_context,
                        BuildConfig.APPLICATION_ID +
                                ".fileProvider", fsms);
                iis.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            }
            else
                uri = Uri.fromFile(fsms);

            iis.putExtra(Intent.EXTRA_STREAM, uri);
            iis.setType("image/png"); // 设置类型
            startActivity(iis);
        });

        binding.btnPlay.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                try {
                    PermissionHelper.checkPermission(this, _permissions);
                    openSDCardFile("/Intent/ring.wav");
                }catch (Exception ex){
                    Log.i(tag,ex.getMessage());
                }
            }
        });

        // 获取SD卡下所有音频文件,然后播放第一首
        binding.btnPlayMediaStore.setOnClickListener(v -> {
            Uri upms = Uri.withAppendedPath(MediaStore.Audio.Media.INTERNAL_CONTENT_URI,
                    "1");
            Intent ipms = new Intent(Intent.ACTION_VIEW, upms);
            startActivity(ipms);
        });

        binding.btnInstallApk.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                try {
                    PermissionHelper.checkPermission(this, _permissions);
                    openSDCardFile("/Download/ui.apk");
                }catch (Exception ex){
                    Log.i(tag,ex.getMessage());
                }
            }
        });

        // 打开另一程序
        binding.btnOpenApp.setOnClickListener(v -> {
            Intent intent = new Intent();
            ComponentName cn = new ComponentName("com.water.uisample",
                    "com.water.uisample.ExpandableListViewActivity");
            intent.setComponent(cn);
            intent.setAction(Intent.ACTION_PACKAGE_FIRST_LAUNCH);
            intent.setAction(Intent.ACTION_MAIN);
            startActivity(intent);
        });

        binding.btnOpenRecord.setOnClickListener(v -> {
            Intent ior = new Intent(MediaStore.Audio.Media.RECORD_SOUND_ACTION);

            if(ior.resolveActivity(getPackageManager())!=null){
                startActivity(ior);
            }else{
                Toast.makeText(_context,"未找到对应应用",Toast.LENGTH_SHORT).show();
            }
        });

        binding.btnTakePhoto.setOnClickListener(v -> {
            Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);

            TakePhoto.launch(intent);
        });

        // 共享数据
        MyApplication ma = MyApplication.getInstance();
        ma.setName("APPLICATION 数据传递");
    }

    private void openSDCardFile(String filename) {
        File file = new File(Environment.getExternalStorageDirectory(),
                filename);

        if(!file.exists())
            return;

        Intent in = new Intent(Intent.ACTION_DEFAULT);
        in.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        Uri uri;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            uri= FileProvider.getUriForFile(this,
                    BuildConfig.APPLICATION_ID +
                            ".fileProvider", file);
            in.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        }
        else
            uri = Uri.fromFile(file);

        if (filename.contains(".doc") || filename.contains(".docx")) {
            in.setDataAndType(uri, "application/msword");
        } else if(filename.contains(".pdf")) {
            in.setDataAndType(uri, "application/pdf");
        } else if(filename.contains(".ppt") || filename.contains(".pptx")) {
            in.setDataAndType(uri, "application/vnd.ms-powerpoint");
        } else if(filename.contains(".xls") || filename.contains(".xlsx")) {
            in.setDataAndType(uri, "application/vnd.ms-excel");
        } else if(filename.contains(".zip") || filename.contains(".rar")) {
            in.setDataAndType(uri, "application/x-wav");
        } else if(filename.contains(".rtf")) {
            in.setDataAndType(uri, "application/rtf");
        } else if(filename.contains(".wav") || filename.contains(".mp3")) {
            in.setDataAndType(uri, "audio/x-wav");
        } else if(filename.contains(".gif")) {
            in.setDataAndType(uri, "image/gif");
        } else if(filename.contains(".jpg") || filename.contains(".jpeg") ||
                filename.contains(".png")) {
            in.setDataAndType(uri, "image/jpeg");
        } else if(filename.contains(".txt")) {
            in.setDataAndType(uri, "text/plain");
        } else if(filename.contains(".3gp") || filename.contains(".mpg") ||
                filename.contains(".mpeg") || filename.contains(".mpe") ||
                filename.contains(".mp4") || filename.contains(".avi")) {
            in.setDataAndType(uri, "video/*");
        }else if (filename.contains(".apk")){
            in.setDataAndType(uri, "application/vnd.android.package-archive");
        } else {
            in.setDataAndType(uri, "*/*");
        }

        startActivity(in);
    }
}

File: app/src/main/java/com/github/androidhappyclub/intentsample/IntentActivity.kt
/*
 * MIT License
 *
 * Copyright (c) 2023 AndroidHappyClub
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.github.androidhappyclub.intentsample

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.ComponentName
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.provider.ContactsContract
import android.provider.MediaStore
import android.provider.Settings
import android.util.Log
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import com.ave.vastgui.tools.activity.result.contract.TakePhotoContract
import com.ave.vastgui.tools.manager.filemgr.FileMgr
import com.ave.vastgui.tools.utils.AppUtils
import com.ave.vastgui.tools.utils.IntentUtils
import com.ave.vastgui.tools.utils.download.DLManager
import com.ave.vastgui.tools.utils.permission.Permission
import com.ave.vastgui.tools.utils.permission.requestPermission
import com.ave.vastgui.tools.view.toast.SimpleToast
import com.ave.vastgui.tools.viewbinding.viewBinding
import com.github.androidhappyclub.intentsample.adapter.IntentItemAdapter
import com.github.androidhappyclub.intentsample.databinding.ActivityIntentBinding
import com.github.androidhappyclub.intentsample.logger.mLogFactory
import java.io.File

class IntentActivity : AppCompatActivity(R.layout.activity_intent) {

    companion object {
        const val FROM_EARTH_KEY = "FROM_EARTH_KEY"
    }

    private val binding: ActivityIntentBinding by viewBinding(ActivityIntentBinding::bind)
    private val logger = mLogFactory.getLog(IntentActivity::class.java)

    /**
     * 注册返回数据回调。
     */
    private val getReturnString =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
            if (result.resultCode == RESULT_OK) {
                val msg = result.data?.extras?.getString(MarsActivity.MarsResultName)
                SimpleToast.showShortMsg(msg.toString())
            }
        }

    /**
     * 注册 [TakePhotoContract] 回调。
     */
    private val takePhoto =
        registerForActivityResult(TakePhotoContract()) { uri ->
            if (null == uri) return@registerForActivityResult
            contentResolver.openInputStream(uri).use {
                val bmp = BitmapFactory.decodeStream(it)
                // 根据旋转角度，生成旋转矩阵
                val matrix = Matrix()
                matrix.postRotate(90f)
                // 将原始图片按照旋转矩阵进行旋转，并得到新的图片
                val bitmap = Bitmap.createBitmap(bmp, 0, 0, bmp.width, bmp.height, matrix, true)
                binding.ivPhoto.setImageBitmap(bitmap)
                bmp.recycle()
            }
        }

    private val recorder =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            logger.d("从录音机中返回")
        }

    private val installRequest =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            if (it.resultCode == Activity.RESULT_OK) {
                installApk()
            }
        }

    private val iiAdapter = IntentItemAdapter(this)

    /**
     * 安装 apk 的根路径。
     */
    private val apkRoot: File by lazy {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            File(filesDir, "Download")
        } else {
            Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
        }
    }

    /**
     * 测试的 wav 文件存放路径。
     */
    private val wavRoot: File by lazy {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            File(filesDir, "Music")
        } else {
            Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC)
        }
    }

    @SuppressLint("MissingPermission")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        downloadApk()
        readWav()

        iiAdapter.addItem("返回首页") {
            val home = Intent().apply {
                //为Intent设置Action和Category属性
                action = Intent.ACTION_MAIN
                addCategory(Intent.CATEGORY_HOME)
            }

            // 使用resolveActivity进行检查，这是十分必要的
            if (home.resolveActivity(packageManager) != null) {
                startActivity(home)
            } else {
                SimpleToast.showShortMsg("未找到对应应用")
            }
        }

        // 给移动客服10086拨打电话
        // 会申请 Manifest.permission.CALL_PHONE 权限
        iiAdapter.addItem("拨打电话") {
            requestPermission(Manifest.permission.CALL_PHONE) {
                granted = {
                    SimpleToast.showShortMsg("拨打电话")
                    IntentUtils.dialPhoneNumber(this@IntentActivity, "10086")
                }
            }
        }

        // 给10086发送内容为“Hello”的短信
        iiAdapter.addItem("发送短信") {
            // 关于smsto的内容，请参考下列链接
            // https://developer.android.com/guide/components/intents-common?hl=zh-cn
            IntentUtils.sendMmsMessage(this, "Hello", "10086")
        }

        // 打开浏览器
        iiAdapter.addItem("打开浏览器") {
            IntentUtils.openWebPage(this, "http://www.baidu.com")
        }

        iiAdapter.addItem("无线设置") {
            val intent = Intent(Settings.ACTION_WIRELESS_SETTINGS)
            if (intent.resolveActivity(packageManager) != null) {
                startActivity(intent)
            } else {
                SimpleToast.showShortMsg("未找到对应应用")
            }
        }

        iiAdapter.addItem("联系人") {
            val intent = Intent().apply {
                action = Intent.ACTION_VIEW
                data = ContactsContract.Contacts.CONTENT_URI
            }
            if (intent.resolveActivity(packageManager) != null) {
                startActivity(intent)
            } else {
                SimpleToast.showShortMsg("未找到对应应用")
            }
        }

        iiAdapter.addItem("显示启动") {
            val intent = Intent(this@IntentActivity, FirstActivity::class.java)
            if (intent.resolveActivity(packageManager) != null) {
                startActivity(intent)
            } else {
                SimpleToast.showShortMsg("未找到对应应用")
            }
        }

        iiAdapter.addItem("返回数据") {
            val mars = Intent(this@IntentActivity, MarsActivity::class.java).apply {
                putExtra(FROM_EARTH_KEY, "地球来的消息:我是来自地球上的Tom，火星的朋友你好!")
            }
            getReturnString.launch(mars)
        }

        iiAdapter.addItem("复杂数据传递") {
            val intent = Intent(this@IntentActivity, ComplexDataActivity::class.java)

            // 传递基本数据类型或String类型数组
            val bd = Bundle()
            bd.putStringArray("StringArray", arrayOf("请求", "全球"))

            // 传递基本数据类型或String类型集合
            intent.putStringArrayListExtra("StringList", object : ArrayList<String>() {
                init {
                    add("请求")
                    add("全球")
                }
            })

            // 传递Bitmap
            val inputStream = resources.assets.open("car.png")
            val bmp = BitmapFactory.decodeStream(inputStream)
            bd.putParcelable("bitmap", bmp)
            intent.putExtras(bd)

            if (intent.resolveActivity(packageManager) != null) {
                startActivity(intent)
            } else {
                SimpleToast.showShortMsg("未找到对应应用")
            }
        }

        // 相当于发送带附件的短信
        iiAdapter.addItem("发送彩信") {
            val picture = File(FileMgr.appInternalFilesDir().path, "/pictures/android.png")
            val uri = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                FileMgr.getFileUriAboveApi24(picture, "${AppUtils.getPackageName()}.fileProvider")
            } else {
                FileMgr.getFileUriOnApi23(picture)
            }
            val intent = Intent(Intent.ACTION_SEND).apply {
                // 短信地址 设置类型
                setDataAndType(Uri.parse("10086"), "image/png")
                // 短信内容
                putExtra("sms_body", "Hello")
                putExtra(Intent.EXTRA_STREAM, uri)
                // 授予 Uri 读取权限
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            }

            if (intent.resolveActivity(packageManager) != null) {
                startActivity(intent)
            } else {
                SimpleToast.showShortMsg("未找到对应应用")
            }
        }

        iiAdapter.addItem("播放文件") {
            requestPermission(Permission.READ_MEDIA_AUDIO) {
                granted = {
                    playWav()
                }
            }
        }

        iiAdapter.addItem("安装APK") {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && !packageManager.canRequestPackageInstalls()) {
                val intent = Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES).apply {
                    data = Uri.parse("package:${AppUtils.getPackageName()}")
                }
                if (intent.resolveActivity(packageManager) != null) {
                    installRequest.launch(intent)
                } else {
                    SimpleToast.showShortMsg("未找到对应应用")
                }
            } else {
                installApk()
            }
        }

        // 打开另一程序
        iiAdapter.addItem("打开程序") {
            val intent = Intent().apply {
                val pkg = "com.github.androidhappyclub.uisample"
                val cn = ComponentName(pkg, "$pkg.ExpandableListViewActivity")
                component = cn
                action = Intent.ACTION_VIEW
            }

            if (intent.resolveActivity(packageManager) != null) {
                startActivity(intent)
            } else {
                SimpleToast.showShortMsg("未找到对应应用")
            }
        }

        iiAdapter.addItem("打开录音机") {
            val intent = Intent(MediaStore.Audio.Media.RECORD_SOUND_ACTION)
            if (intent.resolveActivity(packageManager) != null) {
                recorder.launch(intent)
            } else {
                SimpleToast.showShortMsg("未找到对应应用")
            }
        }

        iiAdapter.addItem("拍照") {
            val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
            takePhoto.launch(intent)
        }

        binding.intents.adapter = iiAdapter
        binding.intents.layoutManager = GridLayoutManager(this, 3)
    }

    /**
     * 下载测试的 apk 文件。
     */
    private fun downloadApk() {
        // 创建根目录
        FileMgr.makeDir(apkRoot).result.onSuccess {
            logger.d("${apkRoot.path} 创建成功")
        }.onFailure {
            logger.e(it.message.toString())
        }

        // 创建下载任务下载测试 apk
        val task = DLManager.createTaskConfig()
            .setDownloadUrl("https://alixiaobai.cn/files/task_stack.apk")
            .setSaveDir(apkRoot.path)
            .setListener {
                onFailure = {
                    Log.e("IntentActivity", "下载遭遇异常 ${it.exception.message}")
                }
                onDownloading = {
                    Log.d("IntentActivity", "当前的下载进度是 ${it.rate}")
                }
            }
            .build()
        task.start()
    }

    /**
     * 安装测试的 apk 文件。
     */
    private fun installApk() {
        val file = File(apkRoot, "task_stack.apk")
        if (!file.exists()) return
        val intent = Intent(Intent.ACTION_DEFAULT)
        val uri: Uri = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            FileMgr.getFileUriAboveApi24(file, "${AppUtils.getPackageName()}.fileProvider")
        } else {
            FileMgr.getFileUriOnApi23(file)
        }
        intent.apply {
            setDataAndType(uri, FileMgr.getMimeType(file, ""))
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        if (intent.resolveActivity(packageManager) != null) {
            startActivity(intent)
        } else {
            SimpleToast.showShortMsg("未找到对应应用")
        }
    }

    /**
     * 从 assets 中读取音频文件
     */
    private fun readWav() {
        val file = FileMgr.getAssetsFile("test.wav", wavRoot)
        if (file.exists()) {
            logger.d("音频文件保存成功")
        } else {
            logger.e("音频文件保存失败")
        }
    }

    private fun playWav() {
        val file = File(wavRoot, "test.wav")
        val uri = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            FileMgr.getFileUriAboveApi24(file, "${AppUtils.getPackageName()}.fileProvider")
        } else {
            FileMgr.getFileUriOnApi23(file)
        }
        val intent = Intent(Intent.ACTION_DEFAULT).apply {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            }
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            setDataAndType(uri, FileMgr.getMimeType(file, ""))
        }
        if (intent.resolveActivity(packageManager) != null) {
            startActivity(intent)
        } else {
            SimpleToast.showShortMsg("未找到对应应用")
        }
    }

}